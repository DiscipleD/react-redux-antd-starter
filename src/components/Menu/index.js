/**
 * Created by jack on 16-9-7.
 */

import React from 'react';
import {Link} from 'react-router';
import {Menu, Icon} from 'antd';

export default React.createClass({
	getInitialState() {
		return this.getKeysFromSelectedItems(this.props.selectedItems)
	},

	componentWillReceiveProps(props) {
		this.setState(this.getKeysFromSelectedItems(props.selectedItems));
	},

	getKeysFromSelectedItems(list = []) {
		const openKeys = [];
		const selectedKeys = [];
		list.forEach(subMenu =>
			subMenu.child.forEach(menuItem => {
				openKeys.push(subMenu.key);
				selectedKeys.push(menuItem.key);
			})
		);
		return {openKeys, selectedKeys};
	},

	toggleSubMenu(e) {
		const openKey = e.key;
		const index = this.state.openKeys.findIndex(key => key === openKey);
		index > -1 ? this.state.openKeys.splice(index, 1) : this.state.openKeys.push(openKey);
	},

	render() {
		console.log('render');
		const styles = {
			height: '100%'
		};
		const {menuList} = this.props;
		const MenuContent = menuList.map(subMenu =>
			<Menu.SubMenu
				key={subMenu.key}
				title={<span><Icon type={subMenu.icon}/>{subMenu.title}</span>}
				onTitleClick={this.toggleSubMenu}
			>
				{subMenu.child.map(menuItem =>
					<Menu.Item key={menuItem.key}>
						<Link to={menuItem.path}>{menuItem.title}</Link>
					</Menu.Item>
				)}
			</Menu.SubMenu>
		);

		return (
			<Menu
				mode="inline"
				selectedKeys={this.state.selectedKeys}
				openKeys={this.state.openKeys}
				style={styles}
			>
				{MenuContent}
			</Menu>
		)
	}
});
