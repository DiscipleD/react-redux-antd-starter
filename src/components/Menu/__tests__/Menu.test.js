/**
 * Created by jack on 16-9-26.
 */

import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import ReactTestUtils from 'react-addons-test-utils';

import Menu from '../index';

describe('Menu Component', () => {
	const menuList = [{
		name: 'user',
		title: '导航一',
		key: 'user',
		icon: 'user',
		child: [{
			name: 'PA',
			title: '项目 A',
			key: 'PA',
			path: '/a'
		}, {
			name: 'PB',
			title: '项目 B',
			key: 'PB',
			path: '/b'
		}]
	}, {
		name: 'notification',
		title: '导航二',
		key: 'notification',
		icon: 'notification',
		child: [{
			name: 'PC',
			title: '项目 C',
			key: 'PC',
			path: '/c'
		}]
	}];
	const selectedItems = [{
		name: 'user',
		title: '导航一',
		key: 'user',
		icon: 'user',
		child: [{
			name: 'PA',
			title: '项目 A',
			key: 'PA',
			path: '/a'
		}]
	}];
	const menu = <Menu menuList={menuList} selectedItems={selectedItems} />;
	const MenuElement = ReactTestUtils.renderIntoDocument(menu);
	const MenuNode = ReactDOM.findDOMNode(MenuElement);

	it('Menu active item', () => {
		expect(MenuNode.querySelectorAll('.ant-menu-item-selected').length).toBe(1);
	});

	it('Menu hidden item', () => {
		// if item doesn't display, init render will not render the subMenu.
		expect(MenuNode.querySelectorAll('.ant-menu-sub').length).toBe(1);
	});

	it('Match Snapshot', () => {
		const snapshot = renderer.create(menu).toJSON();
		expect(snapshot).toMatchSnapshot();
	});
});
