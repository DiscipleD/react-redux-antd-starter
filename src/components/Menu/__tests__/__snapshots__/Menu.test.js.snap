exports[`Menu Component Match Snapshot 1`] = `
<ul
  aria-activedescendant=""
  className="ant-menu ant-menu-inline  ant-menu-light ant-menu-root"
  onKeyDown={[Function bound onKeyDown]}
  role="menu"
  style={
    Object {
      "height": "100%"
    }
  }
  tabIndex="0">
  <li
    className="ant-menu-submenu-inline ant-menu-submenu-open ant-menu-submenu-selected ant-menu-submenu"
    onMouseEnter={[Function bound onMouseEnter]}
    onMouseLeave={[Function bound onMouseLeave]}>
    <div
      aria-expanded={true}
      aria-haspopup="true"
      aria-owns="user$Menu"
      className="ant-menu-submenu-title"
      onClick={[Function bound onTitleClick]}
      onMouseEnter={[Function bound onTitleMouseEnter]}
      onMouseLeave={[Function bound onTitleMouseLeave]}
      style={
        Object {
          "paddingLeft": 24
        }
      }>
      <span>
        <i
          className="anticon anticon-user" />
        导航一
      </span>
    </div>
    <ul
      aria-activedescendant=""
      className="ant-menu ant-menu-inline  ant-menu-sub"
      id="user$Menu"
      role="menu"
      style={Object {}}>
      <li
        aria-disabled={undefined}
        aria-selected={true}
        className="ant-menu-item-selected ant-menu-item"
        onClick={[Function bound onClick]}
        onMouseEnter={[Function bound onMouseEnter]}
        onMouseLeave={[Function bound onMouseLeave]}
        role="menuitem"
        style={
          Object {
            "paddingLeft": 48
          }
        }
        title={undefined}>
        <a
          onClick={[Function bound handleClick]}
          style={Object {}}>
          项目 A
        </a>
      </li>
      <li
        aria-disabled={undefined}
        aria-selected={false}
        className="ant-menu-item"
        onClick={[Function bound onClick]}
        onMouseEnter={[Function bound onMouseEnter]}
        onMouseLeave={[Function bound onMouseLeave]}
        role="menuitem"
        style={
          Object {
            "paddingLeft": 48
          }
        }
        title={undefined}>
        <a
          onClick={[Function bound handleClick]}
          style={Object {}}>
          项目 B
        </a>
      </li>
    </ul>
  </li>
  <li
    className="ant-menu-submenu-inline ant-menu-submenu"
    onMouseEnter={[Function bound onMouseEnter]}
    onMouseLeave={[Function bound onMouseLeave]}>
    <div
      aria-expanded={false}
      aria-haspopup="true"
      aria-owns="notification$Menu"
      className="ant-menu-submenu-title"
      onClick={[Function bound onTitleClick]}
      onMouseEnter={[Function bound onTitleMouseEnter]}
      onMouseLeave={[Function bound onTitleMouseLeave]}
      style={
        Object {
          "paddingLeft": 24
        }
      }>
      <span>
        <i
          className="anticon anticon-notification" />
        导航二
      </span>
    </div>
  </li>
</ul>
`;
